### PR TITLE
Scoped signatures

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/544

I have a problem with makeTransaction, because i need the witness ...

### CODE REVIEW

File: neo/Network/P2P/Payloads/Witness.cs [hunk 1]
```diff
@@ -24,16 +25,29 @@ public virtual UInt160 ScriptHash
             }
         }
 
-        public int Size => InvocationScript.GetVarSize() + VerificationScript.GetVarSize();
+        public bool HasScopedHash => Scope == WitnessScope.CustomScriptHash || Scope == WitnessScope.ExecutingGroupPubKey;
```
Comment (igormcoelho) [inline neo/Network/P2P/Payloads/Witness.cs]: This is very well done Shargon, congratulations. We just need to filter this pubkey here, as it has another size other than Hash160.... I'm not against groupHash concept Hash160(pubkey) or Hash160(21+pubkey+ac concept...), and group becomes a witness :joy: or just keep pubkey, and count bytes correctly ;)

File: neo/Consensus/ConsensusContext.cs [hunk 2]
```diff
@@ -82,7 +82,7 @@ public Block CreateBlock()
             for (int i = 0, j = 0; i < Validators.Length && j < M; i++)
             {
                 if (CommitPayloads[i]?.ConsensusMessage.ViewNumber != ViewNumber) continue;
-                sc.AddSignature(contract, Validators[i], CommitPayloads[i].GetDeserializedMessage<Commit>().Signature);
+                sc.AddSignature(contract, Validators[i], CommitPayloads[i].GetDeserializedMessage<Commit>().Signature, WitnessScope.Global, UInt160.Zero);
```
Comment (igormcoelho) [inline neo/Consensus/ConsensusContext.cs]: Should we keep Global as default param? Same for filter zero?

File: neo/Network/P2P/Payloads/ConsensusPayload.cs [hunk 3]
```diff
@@ -85,6 +85,7 @@ void ISerializable.Deserialize(BinaryReader reader)
             ((IVerifiable)this).DeserializeUnsigned(reader);
             if (reader.ReadByte() != 1) throw new FormatException();
             Witness = reader.ReadSerializable<Witness>();
+            if (Witness.Scope != WitnessScope.Global) throw new FormatException();
```
Comment (igormcoelho) [inline neo/Network/P2P/Payloads/ConsensusPayload.cs]: Very good move. Blocks require global witness.

File: neo/Network/P2P/Payloads/WitnessScope.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        /// <summary>
+        /// (neo2) - no params
+        /// </summary>
+        Global = 0x00,
+
+        /// <summary>
+        /// EntryScriptHash - no params (root-only witnesses) - can be default safe choice for native NEO/GAS (neo2 attach mode)
+        /// </summary>
+        InitScriptHash = 0x01,
```
Comment (igormcoelho) [inline neo/Network/P2P/Payloads/WitnessScope.cs]: This should ensure that `EntryScriptHash == CallingScriptHash`, during CheckWitness verification.

File: neo/SmartContract/InteropService.cs [hunk 5]
```diff
@@ -143,8 +143,30 @@ private static bool Runtime_GetTrigger(ApplicationEngine engine)
...
+            {
+                if (witness.ScriptHash != hash) continue;
+
+                switch (witness.Scope)
+                {
+                    default: return false;
+
+                    case WitnessScope.Global: return true;
+                    case WitnessScope.CustomScriptHash: return engine.CurrentScriptHash == witness.ScopedHash;
+                    case WitnessScope.InitScriptHash: return engine.CurrentScriptHash == engine.EntryScriptHash;
```
Comment (igormcoelho) [inline neo/SmartContract/InteropService.cs]: Current or Calling?

**Comment by igormcoelho** [comment]
@shargon I may need to change quite a few things here, is that ok?

**Comment by vncoelho** [comment]
Maybe it is better to open a new one I think, Igor.

**Comment by shargon** [comment]
@igormcoelho feel free to modify

**Comment by shargon** [comment]
I will close it, and create a new one cleaner

**Comment by igormcoelho** [comment]
Ok, please start it as clean as possible, so we make sure core functionality works fine and we focus on that.

**Comment by igormcoelho** [comment]
Shargon, I'll try to start another one in a very different direction... this way we increase the odds of finishing this quickly.

**Comment by shargon** [comment]
Is duplicate work

**Comment by igormcoelho** [comment]
> Is duplicate work

why?

**Comment by igormcoelho** [comment]
@shargon I opened another PR, but please cleanup and re-open yours. This way, we can have complementary views that may solve the problem. Or if you prefer, re-open yours and take parts of mine that help solving the issue (if any).

**Pull request closed without merging**